Update Declarative Web Push to some recent standards changes

beidson opened the pull request:
#### 986e05d99d9e6bde8dd22bb035a815564994ce2b
<pre>
Update Declarative Web Push to some recent standards changes
<a href="https://rdar.apple.com/114886376">rdar://114886376</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=284094">https://bugs.webkit.org/show_bug.cgi?id=284094</a>

Reviewed by Ben Nham.

This patch:
- Detects as a declarative push message via the new special JSON key/value
- Shows the notification automatically when possible
- Enables it (retaining the switch to disable)

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/notifications/NotificationJSONParser.cpp:
(WebCore::hasDeclarativeMessageJSONValues):
(WebCore::NotificationJSONParser::hasDeclarativeMessageHeader):
(WebCore::NotificationJSONParser::parseNotificationPayload):
(WebCore::NotificationJSONParser::parseNotificationOptions):
(WebCore::dirKey): Deleted.
(WebCore::langKey): Deleted.
(WebCore::bodyKey): Deleted.
(WebCore::tagKey): Deleted.
(WebCore::iconKey): Deleted.
(WebCore::dataKey): Deleted.
(WebCore::silentKey): Deleted.
(WebCore::defaultActionURLKey): Deleted.
(WebCore::titleKey): Deleted.
(WebCore::appBadgeKey): Deleted.
(WebCore::mutableKey): Deleted.
* Source/WebCore/Modules/notifications/NotificationJSONParser.h:
* Source/WebCore/Modules/notifications/NotificationPayload.cpp:
(WebCore::NotificationPayload::hasDeclarativeMessageHeader):
(WebCore::NotificationPayload::toNotificationData const):
* Source/WebCore/Modules/notifications/NotificationPayload.h:
* Source/WebKit/webpushd/WebPushDaemon.h:
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::platformDefaultActionBundleIdentifier):
(WebPushD::WebPushDaemon::injectPushMessageForTesting):
(WebPushD::supportsBuiltinNotifications):
(WebPushD::WebPushDaemon::handleIncomingPushImpl):
(WebPushD::WebPushDaemon::showNotification):

* Tools/DumpRenderTree/mac/DumpRenderTree.mm:
(setWebPreferencesForTestOptions):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:
(TestWebKitAPI::(WebPushD, DeclarativeWebPushHandling)):
* LayoutTests/platform/mac-wk2/fast/dom/navigator-detached-no-crash-expected.txt:

Canonical link: <a href="https://commits.webkit.org/287485@main">https://commits.webkit.org/287485@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/53d21229d9274f93b6eecc01fe315e8e52b4812c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79833 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58834 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33232 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84358 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30832 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81942 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67908 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7124 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62399 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20241 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82901 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52459 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72705 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42710 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49799 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26854 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29283 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/72842 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70926 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27327 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85784 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/78930 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7064 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/4952 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70663 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7238 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68545 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69905 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17422 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13912 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12827 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/101301 "Built successfully") | 
| | [❌ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7023 "Failed to checkout and rebase branch from PR 37481") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/12595 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24699 "Found 96 jsc stress test failures: stress/json-stringify-inspector-check.js.no-llint, stress/sampling-profiler-anonymous-function.js.bytecode-cache, stress/sampling-profiler-anonymous-function.js.default, stress/sampling-profiler-anonymous-function.js.dfg-eager, stress/sampling-profiler-anonymous-function.js.dfg-eager-no-cjit-validate ..., JSC test binary failure: testapi") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6881 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10393 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8687 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Source/WebKit/webpushd/WebPushDaemon.mm:
```diff
@@ -533,7 +532,36 @@ static void tryCloseRequestConnection(xpc_object_t request)
+            message.notificationPayload = exceptionOrPayload.returnValue();
+    }
+
+#if HAVE(FULL_FEATURED_USER_NOTIFICATIONS)
```
This does not seem right for macOS Safari since we need to defer to SafariNotificationAgent to show the notification using the legacy web notification center.

beidson replied:
Yup, handling this differently.

bnham commented on Source/WebKit/webpushd/WebPushDaemon.mm:
```diff
@@ -125,12 +125,12 @@ static bool platformShouldPlaySound(const WebCore::NotificationData& data)
+    return (NSString *)identifier.bundleIdentifier;
 #else
     // FIXME: Calculate appropriate value on macOS
     return nil;
```
If we want to enable declarative push for macOS Dock apps then this does not seem correct...

beidson replied:
For now this allows us to API test on Mac, which I think is the highest immediate priority.
This will be revisited in the near future when I get to that.

bnham commented on Tools/TestWebKitAPI/Configurations/TestWebKitAPIBase.xcconfig:
```diff
@@ -126,7 +126,7 @@ WK_WRITING_TOOLS_UI_LDFLAGS_MACOS_SINCE_1500 = -weak_framework WritingToolsUI;
 OTHER_CPLUSPLUSFLAGS = $(inherited) -isystem $(SDKROOT)/System/Library/Frameworks/System.framework/PrivateHeaders;
 
-OTHER_LDFLAGS = $(inherited) $(GTEST_LDFLAGS) -lxml2 -force_load $(BUILT_PRODUCTS_DIR)/libTestWebKitAPI.a -framework JavaScriptCore -framework WebCore -framework WebKit -lWebCoreTestSupport -framework Metal -framework IOSurface $(WK_APPSERVERSUPPORT_LDFLAGS) $(WK_AUTHKIT_LDFLAGS) -framework Network -framework UniformTypeIdentifiers -framework CoreFoundation -framework CoreServices -framework CFNe ...
+OTHER_LDFLAGS = $(inherited) $(GTEST_LDFLAGS) -lxml2 -force_load $(BUILT_PRODUCTS_DIR)/libTestWebKitAPI.a -framework JavaScriptCore -framework WebCore -framework WebKit -lWebCoreTestSupport -framework Metal -framework IOSurface $(WK_APPSERVERSUPPORT_LDFLAGS) $(WK_AUTHKIT_LDFLAGS) -framework Network -framework UniformTypeIdentifiers -framework CoreFoundation -framework CoreServices -framework CFNe ...
```
Does TestWebKitAPI actually have to link against UserNotifications? Doesn't seem like it's used.

beidson replied:
Leftovers from when I was doing explicit things for testing, but dropped (see the FIXME in the test)
WIll revert.

bnham replied:
I think you still need to remove this from the latest patch.

bnham commented on Source/WebCore/Modules/notifications/NotificationJSONParser.cpp:
```diff
@@ -111,18 +122,39 @@ static std::optional<unsigned long long> parseAppBadgeString(const String& badge
+    }
+    if (!declarativeWebPushValue)
+        return Exception { ExceptionCode::NotFoundError, makeString("Push message does not have a valid '"_s, declarativeWebPushKey(), "' integer value"_s) };
+    if (*declarativeWebPushValue != declarativeWebPushValue)
```
Is the check on L132 sufficient? Not totally understanding the point of this check

beidson replied:
The standard behavior for opting in to declarative web push is to have a magic key/value in the push message JSON.
That's what this check is for.

bnham replied:
But doesn't your declaration on L76 shadow the constant declared in L37? It seems like a non-sensical comparison to compare a value against itself.

bnham commented on Source/WebCore/Modules/notifications/NotificationJSONParser.cpp:
```diff
@@ -34,6 +34,22 @@
 namespace WebCore {
 
+static constexpr int declarativeWebPushValue = 8030;
+static ASCIILiteral declarativeWebPushKey()
```
nit: should all of these just be:

```
static constexpr ASCIILiteral declarativeWebPushKey { "web_push"_s };
static constexpr ASCIILiteral notificationKey { "notification"_s };
...
```

beidson replied:
That's probably the new way, yes!

cdumez replied:
Darin would even encourage `auto` here:
```
static constexpr auto declarativeWebPushKey = "web_push"_s;
```

beidson replied:
I updated to Ben's suggestion, but will not be following Darin's suggestion at this time 😁

cdumez commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -2435,19 +2435,19 @@ DeclarativeShadowRootsSerializerAPIsEnabled:
   defaultValue:
     WebKitLegacy:
-      default: false
+      default: true
```
Surprised we support declarative Web Push (and Web Push in general) in WebKitLegacy.

beidson replied:
It's not. The key just exists there because of established momentum of always listing all 3 *shrug*

cdumez commented on Source/WebCore/Modules/notifications/NotificationPayload.h:
```diff
@@ -59,6 +59,8 @@ struct NotificationPayload {
     WEBCORE_EXPORT static ExceptionOr<NotificationPayload> parseJSON(const String&);
     NotificationPayload static fromNotificationData(const NotificationData&);
 
+    WEBCORE_EXPORT NotificationData toNotificationData();
```
`WEBCORE_EXPORT NotificationData toNotificationData() const;`

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:
```diff
@@ -2381,29 +2384,35 @@ static size_t expectedSuccessfulMessages()
     TestWebKitAPI::Util::run(&done);
 
+#if HAVE(FULL_FEATURED_USER_NOTIFICATIONS)
+    auto configuration = adoptNS([[_WKWebPushDaemonConnectionConfiguration alloc] init]);
```
Would be nice to not hide the smart pointers behind `auto` in new code and use `RetainPtr configuration ` here.

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:
```diff
@@ -2381,29 +2384,35 @@ static size_t expectedSuccessfulMessages()
+    configuration.get().bundleIdentifierOverrideForTesting = @"com.apple.WebKit.TestWebKitAPI";
+    configuration.get().hostApplicationAuditToken = getSelfAuditToken();
+    configuration.get().partition = @"TestWebKitAPI";
+    auto connection = adoptNS([[_WKWebPushDaemonConnection alloc] initWithConfiguration:configuration.get()]);
```
ditto.

bnham commented on Source/WebCore/Modules/notifications/NotificationJSONParser.cpp:
```diff
@@ -111,41 +71,62 @@ static std::optional<unsigned long long> parseAppBadgeString(const String& badge
+    if (notificationObject->type() != JSON::Value::Type::Object)
+        return Exception { ExceptionCode::SyntaxError, makeString("Push message '"_s, notificationKey, "' value is not the expected object type"_s) };
+
+    auto protectedObject = notificationObject->asObject();
```
nit: `RefPtr protectedObject = ...`

annevk replied:
This feedback still applies it seems.

bnham reviewed (approved):
r+ pending EWS passing

annevk commented on Source/WebCore/Modules/notifications/NotificationJSONParser.cpp:
```diff
@@ -34,60 +34,20 @@
+static constexpr ASCIILiteral dirKey { "dir"_s };
+static constexpr ASCIILiteral langKey { "lang"_s };
+static constexpr ASCIILiteral bodyKey { "body"_s };
+static constexpr ASCIILiteral navigationURLKey { "navigation"_s };
```
```suggestion
static constexpr ASCIILiteral navigationURLKey { "navigate"_s };
```

annevk commented on Source/WebCore/Modules/notifications/NotificationJSONParser.cpp:
```diff
@@ -111,41 +71,62 @@ static std::optional<unsigned long long> parseAppBadgeString(const String& badge
+            return Exception { ExceptionCode::SyntaxError, makeString("Push message with Notification disposition: '"_s, navigationURLKey, "' member is specified but is not a string"_s) };
 
-        defaultActionURL = URL { value->asString() };
+        navigationURL = URL { value->asString() };
```
The specification uses the scope URL as the base URL.

beidson replied:
> I'm also not seeing the changes here to the Notification object and to PushEvent (we're using the push event now and not pushnotification. I think we should have those changes before we enable by default.

These were missed due to not having an easy way to look at the current spec changes.
They're trivial to implement, but move this patch beyond "small changes since reviewed"

So I'll land this and followup.

annevk reviewed (changes requested):
This looks okay, but I noticed one somewhat critical mistake and another that would be good to fix.

I'm also not seeing the changes here to the Notification object and to PushEvent (we're using the `push` event now and not `pushnotification`. I think we should have those changes before we enable by default.

beidson committed:
Update Declarative Web Push to some recent standards changes
rdar://114886376
https://bugs.webkit.org/show_bug.cgi?id=284094

Reviewed by Ben Nham.

This patch:
- Detects as a declarative push message via the new special JSON key/value
- Shows the notification automatically when possible
- Enables it (retaining the switch to disable)

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/notifications/NotificationJSONParser.cpp:
(WebCore::hasDeclarativeMessageJSONValues):
(WebCore::NotificationJSONParser::hasDeclarativeMessageHeader):
(WebCore::NotificationJSONParser::parseNotificationPayload):
(WebCore::NotificationJSONParser::parseNotificationOptions):
(WebCore::dirKey): Deleted.
(WebCore::langKey): Deleted.
(WebCore::bodyKey): Deleted.
(WebCore::tagKey): Deleted.
(WebCore::iconKey): Deleted.
(WebCore::dataKey): Deleted.
(WebCore::silentKey): Deleted.
(WebCore::defaultActionURLKey): Deleted.
(WebCore::titleKey): Deleted.
(WebCore::appBadgeKey): Deleted.
(WebCore::mutableKey): Deleted.
* Source/WebCore/Modules/notifications/NotificationJSONParser.h:
* Source/WebCore/Modules/notifications/NotificationPayload.cpp:
(WebCore::NotificationPayload::hasDeclarativeMessageHeader):
(WebCore::NotificationPayload::toNotificationData const):
* Source/WebCore/Modules/notifications/NotificationPayload.h:
* Source/WebKit/webpushd/WebPushDaemon.h:
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::platformDefaultActionBundleIdentifier):
(WebPushD::WebPushDaemon::injectPushMessageForTesting):
(WebPushD::supportsBuiltinNotifications):
(WebPushD::WebPushDaemon::handleIncomingPushImpl):
(WebPushD::WebPushDaemon::showNotification):

* Tools/DumpRenderTree/mac/DumpRenderTree.mm:
(setWebPreferencesForTestOptions):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:
(TestWebKitAPI::(WebPushD, DeclarativeWebPushHandling)):
* LayoutTests/platform/mac-wk2/fast/dom/navigator-detached-no-crash-expected.txt:

Canonical link: https://commits.webkit.org/287485@main
